packages/utils/src/internals/error_snapshotter.ts

```diff
@@ -0,0 +1,100 @@
+        // If the key-value store is not available, or the body and page are not available, return empty filenames
+        if (!keyValueStore || (!body && !page)) {
+            return {};
+        }
+
+        const filename = this.generateFilename(error);
+        const screenshotFilename = page ? await this.captureScreenshot(page, keyValueStore, filename) : undefined;
+
+        let htmlFilename: string | undefined;
+        if (typeof body === 'string') {
+            htmlFilename = await this.saveHTMLSnapshot(body, keyValueStore, filename);
+        } else if (page) {
+            const html = await page?.content() || '';
+            htmlFilename = html ? await this.saveHTMLSnapshot(html, keyValueStore, filename) : '';
+        }
+
+        const basePath = `${ErrorSnapshotter.KEY_VALUE_STORE_PATH}/${keyValueStore.id}/records`;
+
+        return {
+            screenshotFilename: screenshotFilename ? `${basePath}/${screenshotFilename}` : undefined,
+            htmlFilename: htmlFilename ? `${basePath}/${htmlFilename}` : undefined,
+        };
+    }
+
+    /**
+     * Capture a screenshot of the page (For Browser only), and return the filename with the extension.
+     */
+    async captureScreenshot(page: PuppeteerPage | PlaywrightPage, keyValueStore: KeyValueStore, filename: string): Promise<string | undefined> {
+        try {
+            const screenshotBuffer = await page.screenshot({ fullPage: true, type: 'png' });
+
+            await keyValueStore.setValue(filename, screenshotBuffer, { contentType: 'image/png' });
+            return `${filename}.png`;
+        } catch (e) {
+            return undefined;
+        }
+    }
+
+    /**
+     * Save the HTML snapshot of the page, and return the filename with the extension.
+     */
+    async saveHTMLSnapshot(html: string, keyValueStore: KeyValueStore, filename: string): Promise<string | undefined> {
+        try {
+            await keyValueStore.setValue(filename, html, { contentType: 'text/html' });
+            return `${filename}.html`;
+        } catch (e) {
+            return undefined;
+        }
+    }
+
+    /**
+     * Generate a unique filename for each error snapshot.
+     */
+    // Method to generate a unique filename for each error snapshot
+    generateFilename(error: ErrnoException): string {
+        // Create a hash of the error stack trace
+        const errorStackHash = crypto.createHash('sha1').update(error.stack || '').digest('hex').slice(0, 30);
+        // Extract the first 30 characters of the error message
+        const errorMessagePrefix = (
+            error.message || ErrorSnapshotter.BASE_MESSAGE
+        ).substring(0, Math.min(ErrorSnapshotter.MAX_ERROR_CHARACTERS, error.message?.length || 0));
+
+        // Generate filename and remove disallowed characters
+        let filename = `${ErrorSnapshotter.SNAPSHOT_PREFIX}_${errorStackHash}_${errorMessagePrefix}`;
+        filename = filename.replace(/[^a-zA-Z0-9!-_.]/g, '');
```

Comment:
I usually replace with `-` so e.g. whitespace are not squashed into single words